crates/cow-amm/src/registry.rs

```diff
@@ -0,0 +1,111 @@
+use {
+    crate::{event_updater::EventUpdater, ContractHandler, CowAmm},
+    ethcontract::common::DeploymentInformation,
+    ethrpc::current_block::{BlockRetrieving, CurrentBlockStream, RangeInclusive},
+    shared::event_handling::{EventRetrieving, EventStoring},
+    std::{collections::BTreeMap, sync::Arc},
+    tokio::sync::RwLock,
+};
+
+/// CoW AMM indexer which stores events in-memory.
+#[derive(Clone)]
+pub struct Registry {
+    cow_amms: Arc<RwLock<BTreeMap<u64, Arc<dyn CowAmm>>>>,
+    first_block: u64,
+}
+
+impl Registry {
+    pub async fn build<W>(
+        block_retriever: Arc<dyn BlockRetrieving>,
+        contract: W,
+        current_block_stream: CurrentBlockStream,
+        deployment_information: Option<DeploymentInformation>,
```

Comment:
A bit annoying that the deployment info is technically already part of `contract`. Since this start block is only relevant in the autopilot (local e2e tests can just set this to 0) I think it's fine to directly take the start block as the argument and lift this logic into the autopilot init setup code.

Reply:
I don't know what to do here, all solutions seems ugly to be honest :sweat_smile: